---
title: "class 15"
author: "Gaeun Jun (A16814573)"
format: pdf
---

### Investigating pertussis cases by year

> Q1. With the help of the R “addin” package datapasta assign the CDC pertussis case number data to a data frame called cdc and use ggplot to make a plot of cases numbers over time.

```{r}
library(datapasta)
library(ggplot2)
```

```{r}
cdc <- data.frame(
  Year = c(1922, 1923,1924,1925,1926,1927,1928,
         1929,1930,1931,1932,1933,1934,1935,
         1936,1937,1938,1939,1940,1941,
         1942,1943,1944,1945,1946,1947,1948,
         1949,1950,1951,1952,1953,1954,
         1955,1956,1957,1958,1959,1960,
         1961,1962,1963,1964,1965,1966,1967,
         1968,1969,1970,1971,1972,1973,
         1974,1975,1976,1977,1978,1979,1980,
         1981,1982,1983,1984,1985,1986,
         1987,1988,1989,1990,1991,1992,1993,
         1994,1995,1996,1997,1998,1999,
         2000,2001,2002,2003,2004,2005,
         2006,2007,2008,2009,2010,2011,2012,
         2013,2014,2015,2016,2017,2018,
         2019,2020,2021,2022),
No..Reported.Pertussis.Cases = c(107473, 164191,165418,152003,202210,181411,
161799,197371,166914,172559,215343,179135,
265269,180518,147237,214652,227319,103188,
183866,222202,191383,191890,109873,
133792,109860,156517,74715,69479,120718,
68687,45030,37129,60886,62786,31732,28295,
32148,40005,14809,11468,17749,17135,
13005,6799,7717,9718,4810,3285,4249,
3036,3287,1759,2402,1738,1010,2177,2063,
1623,1730,1248,1895,2463,2276,3589,
4195,2823,3450,4157,4570,2719,4083,6586,
4617,5137,7796,6564,7405,7298,7867,
7580,9771,11647,25827,25616,15632,10454,
13278,16858,27550,18719,48277,28639,
32971,20762,17972,18975,15609,18617,6124,
2116,3044)
)

colnames(cdc) <- c("Year", "Cases")
```

```{r}
ggplot(cdc) +
  aes(x = Year, y = Cases) +
  geom_point(color = "black") +
  geom_line(color = "black") +
  labs(
    title = "Pertussis Cases Over Time",
    x = "Year",
    y = "Number of Cases"
  )
```

### 2. A tale of two vaccines (wP & aP)

> Q2. Using the ggplot geom_vline() function add lines to your previous plot for the 1946 introduction of the wP vaccine and the 1996 switch to aP vaccine (see example in the hint below). What do you notice?

```{r}
ggplot(cdc) +
  aes(x = Year, y = Cases) +
  geom_point(color = "black") +
  geom_line(color = "black") +
  geom_vline(xintercept = 1946, linetype = "dashed", color = "blue") +
  geom_vline(xintercept = 1996, linetype = "dashed", color = "red") +
  labs(
    title = "Pertussis Cases Over Time",
    x = "Year",
    y = "Number of Cases"
  )
```

I noticed that the aP vaccine got introduced after the Pertussis cases have been declining (even with the wP vaccine).

> Q3. Describe what happened after the introduction of the aP vaccine? Do you have a possible explanation for the observed trend?

After the introduction of the aP vaccine, the the number of cases stayed pretty consistently low. This trend could possibly be explained by the fact that this new vaccine has antigens that target the right bacteria. However, after a little bit, the cases start to rise again and this could be due to a growing immunization and resistance to the vaccine administration. 

### 3. Exploring CMI-PB data

##### The CMI-PB API returns JSON data

```{r}
library(jsonlite)
```

```{r}
subject <- read_json("https://www.cmi-pb.org/api/subject", simplifyVector = TRUE) 
```

```{r}
head(subject, 3)
```

> Q4. How many aP and wP infancy vaccinated subjects are in the dataset?

```{r}
table(subject$infancy_vac)
```

There are 60 aP and 85 wP infancy vaccianted subjects in the dataset. 

> Q5. How many Male and Female subjects/patients are in the dataset?

```{r}
table(subject$biological_sex)
```

There are 112 females and 60 males in the dataset. 

> Q6. What is the breakdown of race and biological sex (e.g. number of Asian females, White males etc…)?

```{r}
table(subject$race, subject$biological_sex) 
```

##### Side-Note: Working with dates

```{r}
library(lubridate)
```

```{r}
today()
```

```{r}
# How many days have passed since new year 2000
today() - ymd("2000-01-01")
```

```{r}
# Now in years
time_length( today() - ymd("2000-01-01"),  "years")
```

> Q7. Using this approach determine (i) the average age of wP individuals, (ii) the average age of aP individuals; and (iii) are they significantly different?

```{r}
# Use today's date to calculate age in days
subject$age <- today() - ymd(subject$year_of_birth)
subject$age_years <- time_length(subject$age, "years")
```

```{r}
library(dplyr)

ap <- subject %>% filter(infancy_vac == "aP")
wp <- subject %>% filter(infancy_vac == "wP")
```

```{r}
round( summary( time_length( ap$age, "years" ) ) )
round( summary( time_length( wp$age, "years" ) ) )
```

```{r}
t.test(ap$age_years, wp$age_years)
```

Average age of wP individuals: 36 years
Average age of aP individuals: 27 years
Since the p-value is less than 0.05, the average ages of the wP and aP individuals are significantly different. 

> Q8. Determine the age of all individuals at time of boost?

```{r}
int <- ymd(subject$date_of_boost) - ymd(subject$year_of_birth)
age_at_boost <- time_length(int, "year")
head(age_at_boost)
```

> Q9. With the help of a faceted boxplot or histogram (see below), do you think these two groups are significantly different?

```{r}
ggplot(subject) +
  aes(time_length(age, "year"),
      fill=as.factor(infancy_vac)) +
  geom_histogram(show.legend=FALSE) +
  facet_wrap(vars(infancy_vac), nrow=2) +
  xlab("Age in years")
```

Basedon the graphs alone, the two groups look very different in their ages, however we can run a statistical significant report to confirm the difference. 

```{r}
x <- t.test(time_length( wp$age, "years" ),
       time_length( ap$age, "years" ))

x$p.value
```

##### Joining multiple tables

Reading the specimen and ab_titer tables and storing as dataframes.

```{r}
specimen <- read_json("https://www.cmi-pb.org/api/specimen", simplifyVector = TRUE) 
titer <- read_json("https://www.cmi-pb.org/api/plasma_ab_titer", simplifyVector = TRUE)
```

> Q9. Complete the code to join specimen and subject tables to make a new merged data frame containing all specimen records along with their associated subject details:

```{r}
meta <- left_join(specimen, subject)
```

```{r}
dim(meta)
```

```{r}
head(meta)
```

> Q10. Now using the same procedure join meta with titer data so we can further analyze this data in terms of time of visit aP/wP, male/female etc.

```{r}
abdata <- inner_join(titer, meta)
dim(abdata)
```

> Q11. How many specimens (i.e. entries in abdata) do we have for each isotype?

```{r}
table(abdata$isotype)
```

> Q12. What are the different $dataset values in abdata and what do you notice about the number of rows for the most “recent” dataset?

```{r}
table(abdata$dataset)
```

The difference in these values is that they vary a lot according to year and the most recent dataset has the lowest number of rows. 

### 4. Examine IgG Ab tier levels

```{r}
igg <- abdata %>% filter(isotype == "IgG")
head(igg)
```

> Q13. Complete the following code to make a summary boxplot of Ab titer levels (MFI) for all antigens:

```{r}
ggplot(igg) +
  aes(x = MFI, antigen) +
  geom_boxplot() + 
    xlim(0,75) +
  facet_wrap(vars(visit), nrow=2)
```

> Q14. What antigens show differences in the level of IgG antibody titers recognizing them over time? Why these and not others?

PT, PRN, FIM2/3, and FHA show differences in the level of IgG antibody titers. It is becuase they are administered widely, they can have bigger range of differences compared to the other vaccines.

```{r}
ggplot(igg) +
  aes(MFI_normalised, antigen, col=infancy_vac ) +
  geom_boxplot(show.legend = FALSE) + 
  facet_wrap(vars(visit), nrow=2) +
  xlim(0,75) +
  theme_bw()
```

```{r}
igg %>% filter(visit != 8) %>%
ggplot() +
  aes(MFI_normalised, antigen, col=infancy_vac ) +
  geom_boxplot(show.legend = FALSE) + 
  xlim(0,75) +
  facet_wrap(vars(infancy_vac, visit), nrow=2)
```

> Q15. Filter to pull out only two specific antigens for analysis and create a boxplot for each. You can chose any you like. Below I picked a “control” antigen (“OVA”, that is not in our vaccines) and a clear antigen of interest (“PT”, Pertussis Toxin, one of the key virulence factors produced by the bacterium B. pertussis).

```{r}
filter(igg, antigen=="OVA") %>%
  ggplot() +
  aes(x = infancy_vac, y = MFI, col=infancy_vac) +
  geom_boxplot(show.legend = FALSE) +
  facet_wrap(vars(visit)) +
  theme_bw()
```

```{r}
filter(igg, antigen=="FIM2/3") %>%
  ggplot() +
  aes(MFI_normalised, col=infancy_vac) +
  geom_boxplot(show.legend = FALSE) +
  facet_wrap(vars(visit)) +
  theme_bw()
```

> Q16. What do you notice about these two antigens time courses and the PT data in particular?

The PT levels rise over time while the OVA levels don't show a consistent increase. This is for both the wP and aP subjects of PT. 

> Q17. Do you see any clear different in aP vs. wP responses?

I don't see any clear trends or differences between the aP and wP responses in OVA antigen, however for the PT antigen, the wP group shows slightly higher IgG titers than the aP group. 

Looking at the 2021 dataset IgG PT antigen levels-time-course: 

```{r}
abdata.21 <- abdata %>% filter(dataset == "2021_dataset")

abdata.21 %>% 
  filter(isotype == "IgG",  antigen == "PT") %>%
  ggplot() +
    aes(x=planned_day_relative_to_boost,
        y=MFI_normalised,
        col=infancy_vac,
        group=subject_id) +
    geom_point() +
    geom_line() +
    geom_vline(xintercept=0, linetype="dashed") +
    geom_vline(xintercept=14, linetype="dashed") +
  labs(title="2021 dataset IgG PT",
       subtitle = "Dashed lines indicate day 0 (pre-boost) and 14 (apparent peak levels)")
```

> Q18. Does this trend look similar for the 2020 dataset?

This trend looks similar to the 202o dataset because in this too, the there is a greater peak for the wP than the aP group. 

###. 5. Obtaining CMI-PB RNASeq data

```{r}
url <- "https://www.cmi-pb.org/api/v2/rnaseq?versioned_ensembl_gene_id=eq.ENSG00000211896.7"

rna <- read_json(url, simplifyVector = TRUE) 
```

```{r}
#meta <- inner_join(specimen, subject)
ssrna <- inner_join(rna, meta)
```

> Q19. Make a plot of the time course of gene expression for IGHG1 gene (i.e. a plot of visit vs. tpm).

```{r}
ggplot(ssrna) +
  aes(x = visit, y = tpm, group=subject_id) +
  geom_point() +
  geom_line(alpha=0.2)
```

> Q20.: What do you notice about the expression of this gene (i.e. when is it at it’s maximum level)?

The gene's maximum level happens at visit 4 (of over 6000 tpm).

> Q21. Does this pattern in time match the trend of antibody titer data? If not, why not?

This pattern does match the trend of antobody titer data because both show that after the boost, the data peaks. 

```{r}
ggplot(ssrna) +
  aes(tpm, col=infancy_vac) +
  geom_boxplot() +
  facet_wrap(vars(visit))
```

```{r}
ssrna %>%  
  filter(visit==4) %>% 
  ggplot() +
    aes(tpm, col=infancy_vac) + geom_density() + 
    geom_rug() 
```


